x-pack/metricbeat/module/aws/s3/s3.go

```diff
@@ -0,0 +1,302 @@
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/x-pack/metricbeat/module/aws"
+	"github.com/elastic/beats/x-pack/metricbeat/module/aws/ec2"
+)
+
+var metricsetName = "s3"
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet(aws.ModuleName, metricsetName, New,
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	*aws.MetricSet
+	moduleConfig   *aws.Config
+	awsConfig      *awssdk.Config
+	regionsList    []string
+	durationString string
+	periodInSec    int
+	logger         *logp.Logger
+}
+
+// metricIDNameMap is a translating map between aws s3 module metric name and cloudwatch s3 metric name.
+var metricIDNameMap = map[string][]string{
+	"BucketSizeBytes": {"bucket.size.bytes"},
+	"NumberOfObjects": {"object.count"},
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The aws s3 metricset is beta.")
+	s3Logger := logp.NewLogger(aws.ModuleName)
+
+	moduleConfig := aws.Config{}
+	if err := base.Module().UnpackConfig(&moduleConfig); err != nil {
+		return nil, err
+	}
+
+	if moduleConfig.Period == "" {
+		err := errors.New("period is not set in AWS module config")
+		s3Logger.Error(err)
+	}
+
+	metricSet, err := aws.NewMetricSet(base)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating aws metricset")
+	}
+
+	// Get a list of regions
+	awsConfig := defaults.Config()
+	awsCreds := awssdk.Credentials{
+		AccessKeyID:     moduleConfig.AccessKeyID,
+		SecretAccessKey: moduleConfig.SecretAccessKey,
+	}
+	if moduleConfig.SessionToken != "" {
+		awsCreds.SessionToken = moduleConfig.SessionToken
+	}
+
+	awsConfig.Credentials = awssdk.StaticCredentialsProvider{
+		Value: awsCreds,
+	}
+
+	awsConfig.Region = moduleConfig.DefaultRegion
+
+	awsConfig.Region = moduleConfig.DefaultRegion
+	svcEC2 := ec2sdk.New(awsConfig)
+	regionsList, err := ec2.GetRegions(svcEC2)
+	if err != nil {
+		err = errors.Wrap(err, "GetRegions failed")
+		s3Logger.Error(err.Error())
+	}
+
+	// Calculate duration based on period
+	durationString, periodSec, err := ec2.ConvertPeriodToDuration(moduleConfig.Period)
+	if err != nil {
+		s3Logger.Error(err.Error())
+		return nil, err
+	}
+
+	// Check if period is set to be multiple of 60s or 300s
```

Comment:
For S3 we have the same limitation as for ec2 instances?

Reply:
Does this mean the collection period for S3 stats should be 24h by default as everything else just creates additional cost with no additional data?